### PR TITLE
[15.1] feat(core): support for nested .gitignore and .nxignore

### PR DESCRIPTION
**Pull request by AgentEnder**
## Current Behavior
Only files ignored by the root .gitignore are ignored by the project graph

## Expected Behavior
Any files ignored by git are ignored by the project graph

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/nx/src/config/workspaces.ts [hunk 1]
```diff
@@ -653,36 +654,12 @@ export function globForProjectFiles(
...
-    '**/node_modules',
-    '/dist',
-    ...globsToExclude,
-  ];
 
-  /**
-   * TODO: This utility has been implemented multiple times across the Nx codebase,
-   * discuss whether it should be moved to a shared location.
-   */
   const ig = ignore();
```
Comment (FrozenPandaz) [inline packages/nx/src/config/workspaces.ts:658]: This isn't used.

Comment (FrozenPandaz) [inline packages/nx/src/config/workspaces.ts:658]: We don't need to save the ignore object, and can do it outside of deduplicateProjectFiles

File: packages/nx/src/daemon/server/watcher.ts [hunk 2]
```diff
@@ -14,47 +14,43 @@ import { handleServerProcessTermination } from './shutdown-utils';
 import { Server } from 'net';
 import ignore from 'ignore';
 import { normalizePath } from '../../utils/path';
+import {
+  getIgnoredGlobs,
+  locatedIgnoreFiles,
+} from 'nx/src/utils/ignore-patterns';
```
Comment (FrozenPandaz) [inline packages/nx/src/daemon/server/watcher.ts:20]: bad import

File: packages/nx/src/hasher/node-based-file-hasher.ts [hunk 3]
```diff
@@ -64,19 +64,10 @@ export class NodeBasedFileHasher extends FileHasherBase {
...
-  ig.add(stripIndents`
-      node_modules
-      tmp
-      dist
-      build    
-    `);
-  return ig;
-}
+  const globs = await getIgnoredGlobs({ ig });
+  ig.add(['node_modules', 'tmp', 'dist', 'build']);
```
Comment (FrozenPandaz) [inline packages/nx/src/hasher/node-based-file-hasher.ts:70]: We can remove `node_modules` as its always ignored

File: packages/nx/src/project-graph/file-utils.ts [hunk 4]
```diff
@@ -49,8 +50,7 @@ export function readFileIfExisting(path: string) {
 
 function getIgnoredGlobs() {
   const ig = ignore();
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/file-utils.ts:52]: This is done a lot. Perhaps we should have an private util for this.

File: packages/nx/src/utils/ignore-patterns.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  ig?: Ignore;
+  nxIgnoreOnly?: boolean;
+}
+
+function getGlobOptions(
+  provided?: GetIgnoredGlobsOptions
+): Required<GetIgnoredGlobsOptions> {
+  return {
+    ig: provided?.ig ?? ignore(),
+    nxIgnoreOnly: provided?.nxIgnoreOnly ?? false,
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/ignore-patterns.ts:24]: Rename this to ignore files

File: packages/nx/src/utils/ignore-patterns.ts [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+export interface GetIgnoredGlobsOptions {
+  ig?: Ignore;
+  nxIgnoreOnly?: boolean;
+}
+
+function getGlobOptions(
+  provided?: GetIgnoredGlobsOptions
+): Required<GetIgnoredGlobsOptions> {
+  return {
+    ig: provided?.ig ?? ignore(),
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/ignore-patterns.ts:23]: We can probably get rid of this if we have a private util to create an ignore test function

File: packages/nx/src/daemon/server/watcher.ts [hunk 7]
```diff
@@ -131,14 +124,14 @@ export async function subscribeToWorkspaceChanges(
...
-        .filter(({ path }) => !ignoreObj.ignores(path));
+        .filter(({ path }) => !ignoreObject.ignores(path));
 
       if (nonIgnoredEvents && nonIgnoredEvents.length > 0) {
         cb(null, nonIgnoredEvents);
       }
     },
     {
-      ignore: getIgnoredGlobs(),
+      ignore: ignoredGlobs,
```
[the remark this answers is not in the document]
Comment (AgentEnder) [reply packages/nx/src/daemon/server/watcher.ts]: Yeah, looks like this is possibly currently semi-broken, but working because we double check it.

**Comment by FrozenPandaz** [comment]
Does this fix https://github.com/nrwl/nx/issues/10244?

**Comment by AgentEnder** [comment]
> Does this fix #10244?

Yeah, it would. I'll backport that fix outside of this PR since it is in a bit of perf purgatory / deprioritized.

**Comment by AgentEnder** [comment]
I'm going to close this out - we can revisit as time permits.

**Pull request closed without merging**